File: Refund/Components/DurationText.razor.cs
using Microsoft.AspNetCore.Components;
using Microsoft.JSInterop;

namespace Refund.Components;

/// <summary>
/// Component that displays a time duration relative to a provided timestamp.
/// Automatically updates the display as time progresses (e.g., "5 minutes ago", "just now", etc.).
/// Useful for displaying creation times, update times, or activity timestamps.
/// </summary>
public partial class DurationText : ComponentBase, IAsyncDisposable
{
    /// <summary>
    /// The reference timestamp from which to calculate the duration.
    /// Required parameter.
    /// </summary>
    [Parameter, EditorRequired] 
    public DateTime Timestamp { get; set; }
    
    /// <summary>
    /// JavaScript runtime for interop functionality.
    /// </summary>
    [Inject] 
    public IJSRuntime JSRuntime { get; set; }
    
    /// <summary>
    /// Unique identifier for the element to target with JavaScript.
    /// </summary>
    private string elementId = $"duration-{Guid.NewGuid()}";
    
    /// <summary>
    /// Reference to this component for JavaScript callbacks.
    /// </summary>
    private DotNetObjectReference<DurationText> _objectReference;
    
    /// <summary>
    /// Reference to the imported JavaScript module.
    /// </summary>
    private IJSObjectReference _module;
    
    /// <summary>
    /// Tracks the previously rendered timestamp to detect changes.
    /// </summary>
    private DateTime? _previousTimestamp;
    
    /// <summary>
    /// Indicates whether the JavaScript module has been initialized.
    /// </summary>
    private bool _isModuleInitialized;
    
    /// <summary>
    /// Tracks whether a timestamp update is pending JavaScript module initialization.
    /// </summary>
    private bool _pendingTimestampUpdate;

    /// <summary>
    /// Handles parameter changes by updating the timer when the timestamp changes.
    /// </summary>
    protected override async Task OnParametersSetAsync()
    {
        // Only proceed if timestamp has changed
        if (_previousTimestamp == Timestamp)
            return;

        _previousTimestamp = Timestamp;

        // If module is initialized, update the timer
        // Otherwise, mark for update after module init
        if (_isModuleInitialized)
        {
            await UpdateTimer();
        }
        else
        {
            _pendingTimestampUpdate = true;
        }
    }

    /// <summary>
    /// Initializes the JavaScript module after the component is rendered.
    /// Sets up the timer for automatic updates.
    /// </summary>
    /// <param name="firstRender">Whether this is the first time the component has rendered</param>
    protected override async Task OnAfterRenderAsync(bool firstRender)
    {
        if (firstRender)
        {
            _objectReference = DotNetObjectReference.Create(this);
            _module = await JSRuntime.InvokeAsync<IJSObjectReference>("import", "./_content/Refund/Components/DurationText.razor.js");
            _isModuleInitialized = true;

            // If there was a pending update (timestamp changed before module was ready)
            // or if this is the first render, initialize the timer
            if (_pendingTimestampUpdate || firstRender)
            {
                await UpdateTimer();
                _pendingTimestampUpdate = false;
            }
        }
    }

    /// <summary>
    /// Updates the timer by calling into JavaScript.
    /// This initializes or resets the duration updater with the current timestamp.
    /// </summary>
    private async Task UpdateTimer()
    {
        if (_module != null)
        {
            await _module.InvokeVoidAsync("initializeDurationUpdater", elementId, Timestamp);
        }
    }

    /// <summary>
    /// Cleans up resources when the component is disposed.
    /// Stops JavaScript timer intervals and disposes references.
    /// </summary>
    public async ValueTask DisposeAsync()
    {
        try 
        {
            if (_module != null)
            {
                await _module.InvokeVoidAsync("cleanupDurationUpdater", elementId);
                await _module.DisposeAsync();
            }
            _objectReference?.Dispose();
        }
        catch (JSDisconnectedException)
        {
            // Safely ignore JS disconnection errors during disposal
        }
    }
}

File: Refund/Components/DurationText.razor.js
/**
 * JavaScript module for the DurationText component.
 * Handles real-time duration formatting and updates.
 */

/**
 * Map to track interval IDs for active duration updaters.
 * Key: Element ID, Value: setInterval ID
 */
const intervals = new Map();

/**
 * Formats a duration in milliseconds to a human-readable string.
 * Adaptively selects the most appropriate time units based on the duration.
 * 
 * @param {number} milliseconds - Duration in milliseconds
 * @returns {string} Formatted duration string
 */
function formatDuration(milliseconds) {
    const totalSeconds = Math.floor(milliseconds / 1000);

    const days = Math.floor(totalSeconds / (24 * 3600));
    const hours = Math.floor((totalSeconds % (24 * 3600)) / 3600);
    const minutes = Math.floor((totalSeconds % 3600) / 60);
    const seconds = totalSeconds % 60;

    if (days >= 1) {
        return `${days} d ${hours} h`;
    }

    if (hours >= 1) {
        return `${hours}:${minutes < 10 ? '0' + minutes : minutes}:${seconds < 10 ? '0' + seconds : seconds}`;
    }

    return `${minutes < 10 ? '0' + minutes : minutes}:${seconds < 10 ? '0' + seconds : seconds}`;
}

/**
 * Initializes or resets a duration updater for a specific element.
 * Sets up a timer to update the duration display every second.
 * 
 * @param {string} elementId - ID of the element to update
 * @param {string|Date} timestamp - Starting timestamp
 */
function initializeDurationUpdater(elementId, timestamp) {
    const element = document.getElementById(elementId);
    if (!element) return;

    /**
     * Updates the duration displayed in the element.
     * Calculates the time difference between now and the starting timestamp.
     */
    function updateDuration() {
        const now = new Date().getTime();
        const startTime = new Date(timestamp).getTime();
        const duration = now - startTime;

        element.textContent = formatDuration(duration);
    }

    // Clear any existing interval for this element
    cleanupDurationUpdater(elementId);

    // Initial update
    updateDuration();

    // Update every second
    const intervalId = setInterval(updateDuration, 1000);
    intervals.set(elementId, intervalId);
}

/**
 * Cleans up a duration updater by clearing its interval.
 * 
 * @param {string} elementId - ID of the element whose updater should be cleaned up
 */
function cleanupDurationUpdater(elementId) {
    const intervalId = intervals.get(elementId);
    if (intervalId) {
        clearInterval(intervalId);
        intervals.delete(elementId);
    }
}

export {
    initializeDurationUpdater,
    cleanupDurationUpdater
};